src/attributes/diagnostics.md

```diff
@@ -252,10 +252,125 @@ When used on a function in a trait implementation, the attribute does nothing.
 > let _ = five();
 > ```
 
+# The `non_exhaustive` attribute
+
+The *`non_exhaustive` attribute* is used to indicate that a type will have
+more fields or variants added in the future. It can be applied to
+[`struct`s][struct], [`enum`s][enum] and `enum` variants.
+
+The `non_exhaustive` attribute uses the [_MetaWord_] syntax and thus does not
+take any inputs.
+
+Within the defining crate, types annotated with `non_exhaustive` behave exactly
+the same as if the type were not annotated with `non_exhaustive`:
+
+```rust,ignore
+#[non_exhaustive]
+pub struct Config {
+    pub window_width: u16,
+    pub window_height: u16,
+}
+
+#[non_exhaustive]
+pub enum Error {
+    Message(String),
+    Other,
+}
+
+pub enum Message {
+    #[non_exhaustive] Send { from: u32, to: u32, contents: String },
+    #[non_exhaustive] Reaction(u32),
+    #[non_exhaustive] Quit,
+}
+
+// Non-exhaustive structs can be constructed as normal within the defining crate.
+let config = Config { window_width: 640, window_height: 480 };
+
+// Non-exhaustive structs can be matched on exhaustively within the defining crate.
+if let Ok(Config { window_width, window_height }) = config {
+    // ...
+}
+
+// Non-exhaustive enums can be matched on exhaustively within the defining crate.
+match error {
+    Error::Message(ref s) => { },
+    Error::Other => { },
+}
+
+match message {
+    // Non-exhaustive variants can be matched on exhaustively within the defining crate.
+    Message::Send { from, to, contents } => { },
+    Message::Reaction(type) => { },
+    Message::Quit => { },
+}
+```
+
+In downstream crates, types annotated with `non_exhaustive` have limitations that
+preserve backwards compatibility when new fields or variants are added.
+
+Non-exhaustive types cannot be constructed in downstream crates:
```

Comment:
It would be good to note what "constructed" means here to attempt something a bit more formal by talking about expression forms (e.g. path expressions, struct expressions, call expressions, etc.) as already defined in the reference and what effects `#[non_exhaustive]` has on those. As written, the examples below is largely guide-level and motivational.

Reply:
I've updated the wording here, including references to the expression forms.